pwiz_tools/Skyline/Util/ParsedMolecule.cs

```diff
@@ -204,6 +206,106 @@ public static bool TryParseFormula(string formula, out ParsedMolecule resultMole
+            RegexOptions.Compiled | RegexOptions.CultureInvariant);
+
+        /// <summary>
+        /// Parse the formula and mass offset declarations in a string presumed to describe a chemical formula and/or mass offsets.
+        /// Is aware of simple math, e.g. "C12H5[-1.2/1.21]-C2H[-1.1]" => "C12H5-C2H", 0.1, 0.11
+        /// </summary>
+        /// <param name="formulaAndMasses">input string  e.g. "C12H5", "C12H5[-1.2/1.21]", "[+1.3/1.31]", "1.3/1.31", "C12H5[-1.2/1.21]-C2H[-1.1]"</param>
+        /// <param name="formulaOut">string with any mass modifiers stripped out e.g.  "C12H5[-1.2/1.21]-C2H[-1.1]" ->  "C12H5-C2H" </param>
+        /// <param name="modMassMonoOut">effect of any mono mass modifiers e.g. "C12H5[-1.2/1.21]-C2H[-1.1]" => 0.1 </param>
+        /// <param name="modMassAverageOut">effect of any avg mass modifiers e.g. "C12H5[-1.2/1.21]-C2H[-1.1]" => 0.11</param>
+        private static void ParseFormulaAndMasses(string formulaAndMasses, out string formulaOut, out double? modMassMonoOut, out double? modMassAverageOut)
+        {
+            modMassMonoOut = null;
+            modMassAverageOut = null;
+            formulaOut = formulaAndMasses?.Trim();
+            if (string.IsNullOrEmpty(formulaOut))
+            {
+                return;
+            }
+
+            if (!char.IsUpper(formulaOut[0]) && !MoleculeMassOffset.StringContainsMassOffsetCue(formulaOut))
+            {
+                // No formula or "[+1.2/1.21]" for regex to match - but could be bare "+1.2/1.21"
+                if (TryParseMassOffsetValues(formulaOut, 1.0, out var monoMassParsed, out var averageMassParsed))
+                {
+                    modMassMonoOut = monoMassParsed;
+                    modMassAverageOut = averageMassParsed;
+                    formulaOut = string.Empty;
+                }
+                return;
+            }
+
+            var matches = FormulaAndMassesRegex.Matches(formulaOut);
+            if (matches.Count > 0)
+            {
+                formulaOut = string.Empty;
+                foreach (Match match in matches)
+                {
+                    double sign = match.Groups[1].Value == @"-" ? -1 : 1;
+                    var thisFormula = match.Groups[2].Value;
+                    if (sign < 0 && !string.IsNullOrEmpty(thisFormula))
+                    {
+                        formulaOut += @"-";
+                    }
+                    formulaOut += thisFormula;
+                    var brackets = match.Groups[3].Value;
+                    if (!string.IsNullOrEmpty(brackets))
+                    {
+                        if (!TryParseMassOffsetValues(brackets, sign, out var monoMassParsed, out var averageMassParsed))
+                        {
+                            throw new ArgumentException(string.Format(UtilResources.ParsedMolecule_SplitFormulaAndMasses_Cannot_parse___0___as_a_formula_and_mass_offset, formulaAndMasses));
```

Comment:
I actually wanted something like:

string.Format(UtilResources.ParsedMolecule_SplitFormulaAndMasses_Cannot_parse___0___as_a_mass_offset, brackets)

But maybe it should provide more detail like:

string.Format(UtilResources.ParsedMolecule_SplitFormulaAndMasses_Cannot_parse___0___as_a_mass_offset_in_the_text__1__, brackets, formulaAndMasses)

The code is failing on the "brackets" text. So, it is a good idea to call that out clearly, but I can see how the extra information about the full starting text might be useful. With either of your more general errors, I worry that the use may be confused looking at the formula itself for errors, when there may be none.

Reply:
Good point.  Now it throws this:
![image](https://github.com/ProteoWizard/pwiz/assets/1200761/ddb1afc1-43ff-487a-85fc-d80c95ac025a)